waku/node/waku_node.nim

```diff
@@ -193,6 +195,158 @@ proc connectToNodes*(
+      ): Future[
+          Result[(seq[(WakuMessageHash, Timestamp)], Option[WakuMessageHash]), string]
+      ] {.async: (raises: []), closure.} =
+        let archiveCursor =
+          if cursor.isSome():
+            some(ArchiveCursor(hash: cursor.get()))
+          else:
+            none(ArchiveCursor)
+
+        let query = ArchiveQuery(
+          includeData: true,
+          cursor: archiveCursor,
+          startTime: some(pruneStart),
+          endTime: some(pruneStop),
+          pageSize: 100,
+        )
+
+        let catchable = catch:
+          await node.wakuArchive.findMessages(query)
+
+        let res =
+          if catchable.isErr():
+            return err(catchable.error.msg)
+          else:
+            catchable.get()
+
+        let response = res.valueOr:
+          return err($error)
+
+        let elements = collect(newSeq):
+          for (hash, msg) in response.hashes.zip(response.messages):
+            (hash, msg.timestamp)
+
+        let cursor =
+          if response.cursor.isNone():
+            none(WakuMessageHash)
+          else:
+            some(response.cursor.get().hash)
+
+        return ok((elements, cursor))
+    )
+
+  let transfer: TransferCallback = proc(
+      hashes: seq[WakuMessageHash], peerId: PeerId
+  ): Future[Result[void, string]] {.async: (raises: []), closure.} =
+    var query = StoreQueryRequest()
+    query.includeData = true
+    query.messageHashes = hashes
+    query.paginationLimit = some(uint64(100))
+
+    while true:
+      let queryRes = catch:
+        await node.wakuStoreClient.query(query, peerId)
+
+      let res =
+        if queryRes.isErr():
+          return err(queryRes.error.msg)
+        else:
+          queryRes.get()
+
+      let response = res.valueOr:
+        return err($error)
+
+      query.paginationCursor = response.paginationCursor
+
+      for kv in response.messages:
+        let handleRes = catch:
+          await node.wakuArchive.handleMessage(kv.pubsubTopic.get(), kv.message.get())
+
+        if handleRes.isErr():
+          trace "message transfer failed", error = handleRes.error.msg
+          # Messages can be synced next time since they are not added to storage yet.
+          continue
+
+        node.wakuSync.ingessMessage(kv.pubsubTopic.get(), kv.message.get())
+
+      if query.paginationCursor.isNone():
+        break
+
+    return ok()
+
+  node.wakuSync = WakuSync.new(
+    peerManager = node.peerManager,
+    maxFrameSize = maxFrameSize,
+    syncInterval = syncInterval,
+    relayJitter = relayJitter,
+    pruneCB = prune,
+    transferCB = some(transfer),
+  ).valueOr:
+    return err("Waku sync initialization failed: " & error)
+
+  let catchRes = catch:
+    node.switch.mount(node.wakuSync, protocolMatcher(WakuSyncCodec))
+  if catchRes.isErr():
+    return err("mounting waku sync to the switch failed: " & catchRes.error.msg)
+
+  if node.started:
+    node.wakuSync.start()
+
+  if not node.wakuArchive.isNil():
```

Comment:
Let's add a comment to clarify the purpose of the following block. I just added one but surely you will give a more sensible one :P
```suggestion
  if not node.wakuArchive.isNil():
  ## Let's initialize the btree set that contains the known messages
```

Reply:
I extracted this logic into it's own proc called `initFillStorage`